src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -0,0 +1,575 @@
+            if (string.IsNullOrEmpty(sourcePath))
+            {
+                return;
+            }
+
+            DebugInfoBuilder dIBuilder = this.GetOrCreateDIBuilder(sourcePath);
+
+            // Get the type information for the variable
+            ResolvedType? resType = value.QSharpType;
+            DIType? dIType = resType == null ? null : this.DebugTypeFromQsharpType(resType, dIBuilder)?.DIType;
+
+            // Get the location information for the variable declaration
+            QsLocation? namespaceLocation = this.CurrentNamespaceElementLocation;
+            Position namespaceOffset = namespaceLocation?.Offset ?? Position.Zero;
+            Position absolutePosition = namespaceOffset + this.TotalOffsetFromStatements() + this.TotalOffsetFromExpressions();
+
+            // LLVM Native API has a bug and will crash if we pass in a null dIType
+            // for either CreateLocalVariable or InsertDeclare (https://bugs.llvm.org/show_bug.cgi?id=52459)
+            if (subProgram != null && dIType != null)
+            {
+                if (dIType.Name.Equals(DebugTypeNotSupportedMessage))
+                {
+                    name += " [value display for this type is not currently supported]";
+                }
+
+                DebugPosition debugLocation = ConvertToDebugPosition(absolutePosition);
+
+                // create the debug info for the local variable
+                DILocalVariable dIVar = dIBuilder.CreateLocalVariable(
+                    subProgram,
+                    name,
+                    dIBuilder.CompileUnit?.File,
+                    debugLocation.Line,
+                    dIType,
+                    alwaysPreserve: true,
+                    DebugInfoFlags.None);
+
+                DILocation dILocation = new DILocation(
+                    this.Context,
+                    debugLocation.Line,
+                    debugLocation.Column,
+                    subProgram);
+
+                if (this.sharedState.CurrentBlock != null)
+                {
+                    if (isMutableBinding)
+                    {
+                        // variable is represented as a pointer to the value
+                        Value variable = ((PointerValue)value).Pointer;
+                        dIBuilder.InsertDeclare(
+                        storage: variable,
+                        varInfo: dIVar,
+                        location: dILocation,
+                        insertAtEnd: this.sharedState.CurrentBlock);
+                    }
+                    else
+                    {
+                        // variable is represented as the value itself
+                        Value variable = value.Value;
+                        dIBuilder.InsertValue(
+                            value: variable,
+                            varInfo: dIVar,
+                            location: dILocation,
+                            insertAtEnd: this.sharedState.CurrentBlock);
+                    }
```

Comment:
```suggestion
                    // The variable is represented as a pointer if it's mutable, or its value if it's immutable.
                    var variable = isMutableBinding ? ((PointerValue)value).Pointer : value.Value;
                    dIBuilder.InsertValue(
                        value: variable,
                        varInfo: dIVar,
                        location: dILocation,
                        insertAtEnd: this.sharedState.CurrentBlock);
```

Reply:
I'm changing this back because I didn't look closely enough when I committed this and it the distinction between InsertValue and InsertDeclare is important